tests/containers/kubectl.pm

```diff
@@ -0,0 +1,136 @@
+    validate_script_output("kubectl config current-context", sub { $_ =~ m/default/ });
+    assert_script_run("kubectl config use-context default");
+    assert_script_run("kubectl config set-context --current --namespace=default");
+    assert_script_run("kubectl config set-context localhost --user=root --namespace=default");
+    assert_script_run("kubectl config use-context localhost");
+    assert_script_run("kubectl config use-context default");
+    assert_script_run("kubectl config unset contexts.localhost");
+
+    ## Test jobs
+    record_info('Test 3', 'job and pod test runs');
+    assert_script_run('kubectl create job sayhello --image=busybox:1.28 -- echo "Hello World"');
+    assert_script_run('kubectl create job gimme-date --image=busybox -- date');
+    validate_script_output("kubectl get jobs --no-headers", qr/sayhello/);
+    validate_script_output("kubectl get jobs --no-headers", qr/gimme-date/);
+    assert_script_run('kubectl wait jobs/sayhello --for=condition=complete --timeout=300s', timeout => 330);
+    assert_script_run('kubectl wait jobs/gimme-date --for=condition=complete --timeout=300s', timeout => 330);
+    # Check job output. First get the pod name
+    my $pod = script_output('kubectl get pods -o name --no-headers=true | grep sayhello');
+    script_retry("kubectl logs $pod | grep 'Hello World'", retry => 5, delay => 10);    # collection of the log can sometime take some time
+    ## Apply a custom Deployment, test scaling
+    record_info('Test 4', 'deployment test');
+    assert_script_run('curl -o deployment.yml ' . data_url('containers/kubectl/deployment.yml'));
+    assert_script_run('kubectl apply -f deployment.yml');
+    assert_script_run('kubectl wait deployment nginx-deployment --for condition=available --timeout=300s', timeout => 330);
+    validate_script_output('kubectl describe deployments/nginx-deployment | grep Replicas', sub { $_ =~ m/2 desired.*2 total/ });
+    assert_script_run('kubectl describe deployments/nginx-deployment');
+    # Scale out
+    assert_script_run('kubectl scale --replicas=5 deployments/nginx-deployment');
+    assert_script_run('kubectl wait deployment nginx-deployment --for condition=available --timeout=300s', timeout => 330);
+    validate_script_output('kubectl describe deployments/nginx-deployment | grep Replicas', sub { $_ =~ m/5 desired.*5 total/ });
+    # Scale in
+    assert_script_run('kubectl scale --replicas=2 deployments/nginx-deployment');
+    assert_script_run('kubectl wait deployment nginx-deployment --for condition=available --timeout=300s', timeout => 330);
+    validate_script_output('kubectl describe deployments/nginx-deployment | grep Replicas', sub { $_ =~ m/2 desired.*2 total/ });
+    # Test the port-forwarding and the webserver
+    my $pid = background_script_run('kubectl port-forward deploy/nginx-deployment 8008:80');
+    sleep(10);    # I don't see a better way to wait for the forward to be ready :-(
```

Comment:
You may do `script_retry('curl` as it ends with `0` only when the forwarding works.

Reply:
I tried this at first, but it is not really better:

```
sleep(10); # wait one iteration to not waste resources
script_retry('curl ... ');
```

Arguably this is better because script_retry tries multiple times in case it is not yet ready. The 10 seconds are rather generous, so missing that timeframe should never be an issue. So, do we really gain something from switching to `script_retry` here? WDYT?

Reply:
I am used to try to avoid `sleep()` - that's it. From [documentation](https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/CONTRIBUTING.md#coding-style):
> Avoid sleep(): Do not use sleep() or simulate a sleep-like functionality. This introduces the risk of either unnecessarily wasting time or just shifting race conditions and making them even harder to investigate. You can use sleep but there are only very limited cases where is the best solution (e.g. yield threads in low-level system code). Here you have the classical problem in software design of [Synchronization](https://en.wikipedia.org/wiki/Synchronization_(computer_science)). sleep() should not be used in place of proper synchronization methods. https://blogs.msmvps.com/peterritchie/2007/04/26/thread-sleep-is-a-sign-of-a-poorly-designed-program/ is one of multiple references that explains this: "The original problem is likely a timing/synchronization issue, ignoring it by hiding it with Thread.Sleep is only going to delay the problem and make it occur in random, hard to reproduce ways." And if the problem does not occur at least time is wasted. We are waiting because some condition changes some time ... keyword(s) is/are some time. It is important to know the exact condition that the code should wait for and check for that instead of delaying program execution for an arbitrary amount of time. For openQA tests in general just try to do what you as a human user would also do. You would not look on your watch waiting for exactly 100s before you look at the screen again, right? :)

Reply:
You are right from a principle standpoint, but there is no rule without exception.

The problem that you have in this particular case, is that `sleep` wastes 10 seconds of running test time, which in principle should be unnecessary. In the lack of a better alternative (I.e. a way to wait for this process to complete), it is IMHO still the most economic way of solving the problem at hand.

Here I see no efficient way of ensuring that the background process has completed. `script_retry` **in this case** does not make the situation anyhow better, as it would still wait between failed attempts. You have the exact same situation as with sleep: You wait for 10 seconds, which is in principle not necessary.

In short: `script_retry` **in this case** is just a `sleep` in disguise. I don't see how this would improve the situation at all.

Reply:
We need to implement new argument `sleep_first => 1`  in `script_retry`  :D :D :D

Reply:
Btw, you use `script_retry` here https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/15215/files#diff-301489d174f96744185348e9a3aec18c327ec62bb49fded2463b5a662b731ffcR36

Reply:
Yes, it was done like this to have a proper "wait for things to be ready with timeout". I actually like it, it's a loop with timeout in single line. In this case, are you sure that 10 seconds is enough in ALL the corner cases that can happen in openQA?

Reply:
> In this case, are you sure that 10 seconds is enough in ALL the corner cases that can happen in openQA?

IMHO it is reasonable to assume that 10 seconds is enough to set a simply firewall NAT rule. And you know very well the answer to all questions that contain "All corner cases" and "openQA" in the same sentence :wink: 

I'll change it as it really doesn't matter which way we do it. There is no functional difference in a healthy test run and weird ones would otherwise probably die.